Resend blocks when modified while sending to client

Jeija opened the pull request:
From what I understand, this can happen when the server sends node updates to the client:
- More than 4 nodes change at the same time (server.cpp:888)
- This causes the server to only send nodes that are 5 nodes or less away from the player (server.cpp:911), for any other nodes the corresponding position in `m_blocks_sent` in `RemoteClient` will be removed to mark those blocks for resending.
- `Server::SendBlocks` calls `RemoteClient::GetNextBlocks` which returns among others the block that was just removed from `m_blocks_sent`
- Here is the problem: While the block is transferred to the client, _the block on the server is modified again_. The corresponding position in `m_blocks_sent` is marked for resending again.
- Before the server can send the new blocks, the client answers with `TOSERVER_GOTBLOCKS`, which causes the block in `m_blocks_sent` to be unmarked (=marked as sent = inserted into `m_blocks_sent`):

``` C++
void RemoteClient::GotBlock(v3s16 p)
{
    /** some other stuff here **/
    m_blocks_sent.insert(p);
}
```
- Result: The server thinks the client's copy of the block is up to date when in fact, the block has been modified in the meantime.

You might think this issue is completely hypothetical due to the short time between sending the block and receiving the answer from the client, but it causes major issues when using mods that modify many nodes at the same time, such as mesecons:
#### These are not mesecons bugs, but caused by minetest itself:

You can easily reproduce this issue by building a small mesecons circuit, most annoyingly using pistons. You will get results like these (screenshots were taken in singleplayer):
![screenshot_20160215_164205](https://cloud.githubusercontent.com/assets/1311964/13053163/4b6025ac-d403-11e5-92ca-64365a27d703.png)
![screenshot_20160215_170452](https://cloud.githubusercontent.com/assets/1311964/13053727/49e7c31c-d406-11e5-8627-1cbef34c10ab.png)
And it is not just a matter of waiting a few seconds for the server to resend the block, you have to manually cause a block update in order to see what is really going on.
#### Solution

My patch makes the server remember which blocks were modified, so it can ignore the `TOSERVER_GOTBLOCKS` answers from the client for those blocks that changed since last sending an update.

Jeija committed:
Resend blocks when modified while sending to client

Megaf commented:
Sounds great! I will give it a try and report back later.
Thanks!

est31 commented on src/clientiface.cpp:
```diff
@@ -370,17 +370,24 @@ void RemoteClient::GetNextBlocks (
 
 void RemoteClient::SentBlock(v3s16 p)
 {
+	if(m_blocks_modified.find(p) != m_blocks_modified.end())
```
Space between `if` and (

est31 commented on src/clientiface.cpp:
```diff
@@ -370,17 +370,24 @@ void RemoteClient::GetNextBlocks (
+	if (m_blocks_modified.find(p) == m_blocks_modified.end())
 	{
-		m_excess_gotblocks++;
+		if(m_blocks_sending.find(p) != m_blocks_sending.end())
```
Space between `if` and (

est31 commented on src/clientiface.cpp:
```diff
@@ -370,17 +370,24 @@ void RemoteClient::GetNextBlocks (
-		m_excess_gotblocks++;
+		if(m_blocks_sending.find(p) != m_blocks_sending.end())
+			m_blocks_sending.erase(p);
+		else
```
put the brace on the same line as the else, separed by one space

est31 commented on src/clientiface.cpp:
```diff
@@ -370,17 +370,24 @@ void RemoteClient::GetNextBlocks (
-	if(m_blocks_sending.find(p) != m_blocks_sending.end())
-		m_blocks_sending.erase(p);
-	else
+	if (m_blocks_modified.find(p) == m_blocks_modified.end())
```
brace on same line as ).

est31 commented on src/clientiface.h:
```diff
@@ -396,6 +396,16 @@ class RemoteClient
+		client reports it has received them to account for blocks
+		that are being modified while on the line.
+
+		Key is position, value is dummy
```
No need to replicate the `value is dummy` for `m_blocks_sent`. It is a mistake left over from commit 6a1670dbc31cc0e44178bbd9ad34ff0d5981a060 (the commit replaced map with set but didnt update the comment).

est31 commented:
See comments above. Otherwise looks good.

Jeija committed:
Clientiface: fix some coding style issues

Jeija commented:
Thank you @est31, I have updated my first commit according to your suggestions.
While doing that though, I found many more the style issues in `clientiface.cpp` (actually there are more `if(` than `if (`). So I added a second commit to this PR that would fix all those issues.

So, feel free to merge only the first commit, it contains the style fixes to my code you suggested, the second just contains some more for the rest of the code.

est31 commented:
Its a good idea to have it merged as two commits.

kilbith commented:
@est31 0.4.14 milestone ?

orwell96 commented:
We definitely need this.
I lately recognized a bug with placing locked chests or other types of nodes that are setting up their inventory in after_place_node callback which appears since [my PR](https://github.com/minetest/minetest/commit/e17fbb31d6db3e6d2d9b68f5793ad8b2146f7bcc) regarding m_nothing_to_send_pause_timer.
1. client sends: placing chest
2. server queues mapblock for resending immediately
3. server calls after_place node
4. client answers with gotblocks and server thinks the client already knows about the inventory being created.
Result: The client does not know about the inventory and shows an empty formspec.
Ironically, i initially set m_nothing_to_send_pause_timer to 0.1 instead of 0.

est31 commented:
Okay, nore has +1ed the PR, see IRC log: http://irc.minetest.ru/minetest-dev/2016-03-12#i_4555202

Merged as 089f9bbe817debc5c30f1d845170909f638c9642
only the first commit. @Jeija I couldn't apply the second commit patch, can you rebase it and add it in a separate PR?
